JMMServer/Tasks/AutoAnimeGroupCalculator.cs

```diff
@@ -0,0 +1,594 @@
+                                relation.RelationType = AnimeRelationType.Prequel;
+                                break;
+                            case "sequel":
+                                relation.RelationType = AnimeRelationType.Sequel;
+                                break;
+                            case "alternative setting":
+                                relation.RelationType = AnimeRelationType.AlternativeSetting;
+                                break;
+                            case "alternative version":
+                                relation.RelationType = AnimeRelationType.AlternativeVersion;
+                                break;
+                            case "same setting":
+                                relation.RelationType = AnimeRelationType.SameSetting;
+                                break;
+                            case "character":
+                                relation.RelationType = AnimeRelationType.Character;
+                                break;
+                            default:
+                                relation.RelationType = AnimeRelationType.Other;
+                                break;
+                        }
+
+                        return relation;
+                    })
+                .ToLookup(k => k.FromId);
+
+            return new AutoAnimeGroupCalculator(relationshipMap, exclusions);
+        }
+
+        /// <summary>
+        /// Gets the ID of the anime represents the group containing the specified <see cref="animeId"/>.
+        /// </summary>
+        /// <param name="animeId">The ID of the anime to get the group's anime ID for.</param>
+        /// <returns>The group's representative anime ID. For anime that don't have any suitable relations,
+        /// this will just be <paramref name="animeId"/>.</returns>
+        public int GetGroupAnimeId(int animeId)
+        {
+            int mainAnimeId;
+
+            // Check to see if the group for the specified anime has been previously calculated
+            if (_animeGroupMap.TryGetValue(animeId, out mainAnimeId))
+            {
+                return mainAnimeId;
+            }
+
+            Dictionary<int, RelationNode> nodes;
+            HashSet<RelationEdge> edges;
+
+            if (BuildGroupGraph(animeId, out nodes, out edges))
+            {
+                mainAnimeId = FindSuitableAnimeForGroup(nodes, edges);
+
+                // Remember main anime ID for all anime in the relation graph
+                foreach (int id in nodes.Keys)
+                {
+                    _animeGroupMap[id] = mainAnimeId;
+                }
+            }
+            else // Specified anime doesn't appear to have any relations
+            {
+                _animeGroupMap[animeId] = mainAnimeId = animeId;
+            }
+
+            return mainAnimeId;
+        }
+
+        private static int FindSuitableAnimeForGroup(Dictionary<int, RelationNode> nodes, HashSet<RelationEdge> edges)

```

Comment:
In this, or a different method toggled by the score setting, I'll need to have either score or first air date, determined by setting.